File: chainofresponsibility/exercise/AbstractDataProcessingHandler.java
package chainofresponsibility.exercise;

import chainofresponsibility.MiddlewareHandler;
import chainofresponsibility.Request;

public class AbstractDataProcessingHandler implements DataProcessingChainHandler {

    private DataProcessingChainHandler nextHandler;
    @Override
    public DataProcessingChainHandler setNext(DataProcessingChainHandler dataProcessingChainHandler) {
        nextHandler = dataProcessingChainHandler;
        return dataProcessingChainHandler;
    }

    @Override
    public DataProcessingResponse handle(Data data) {
        if (nextHandler != null) {
            return nextHandler.handle(data);
        }
        return new DataProcessingResponse("Succeeded to proceed the job", true);
    }
}

File: typescript/chain-of-responsibility/exercise/abstract-data-processing-handler.ts
import {Data} from "./data";
import {DataProcessingChainHandler} from "./data-processing-chain-handler";
import {DataProcessingResponse} from "./data-processing-response";

export class AbstractDataProcessingHandler
  implements DataProcessingChainHandler
{
  private nextHandler: DataProcessingChainHandler;

  constructor() {
    this.nextHandler = {
      setNext: () => this,
      handle: () => {
        return {reason: "Succeeded to proceed the job", isSucceeded: true};
      },
    };
  }
  setNext(
    dataProcessingChainHandler: DataProcessingChainHandler
  ): DataProcessingChainHandler {
    this.nextHandler = dataProcessingChainHandler;
    return dataProcessingChainHandler;
  }

  handle(data: Data): DataProcessingResponse {
    if (this.nextHandler != null) {
      return this.nextHandler.handle(data);
    }
    return {reason: "Succeeded to proceed the job", isSucceeded: true};
  }
}
